src/journal/views.py

```diff
@@ -1392,36 +1392,73 @@ def search(request):
 
-    if request.session.get('article_search'):
-        search_term = request.session.get('article_search')
+    if redir:
+        return redir
+    #if no order, set order to title (alphabetical). this happens when receiveing a GET request for keyword.
+    if request.GET and not sort:
+        sort='title'
 
+    if search_term:
+        # return two lists of pks based on search term.
         article_search = submission_models.Article.objects.filter(
             (Q(title__icontains=search_term) |
-             Q(subtitle__icontains=search_term)) &
-            Q(journal=request.journal)
+             Q(keywords__word__icontains=search_term) |
+             Q(subtitle__icontains=search_term)), 
+            journal=request.journal, 
+            stage=submission_models.STAGE_PUBLISHED,
+            date_published__lte=timezone.now()
         )
-        article_search = [article for article in article_search]
+        article_search_unsorted = [article.pk for article in article_search]
 
-        author_search = search_term.split(' ')
         from_author = submission_models.FrozenAuthor.objects.filter(
-            (Q(first_name__in=author_search) |
-             Q(last_name__in=author_search)) &
-            Q(article__journal=request.journal)
+            (Q(first_name__icontains=search_term) |
+             Q(last_name__icontains=search_term)),
+            article__journal=request.journal, 
+            article__stage=submission_models.STAGE_PUBLISHED, 
+            article__date_published__lte=timezone.now()
         )
-
-        articles_from_author = [author.article for author in from_author]
-        articles = set(article_search + articles_from_author)
+        articles_from_author_unsorted = [author.article.pk for author in from_author]
+        # merge the two lists into a set which will clear out duplicates
+        article_list_unsorted = set(article_search_unsorted + articles_from_author_unsorted)
+        # do a search on the set of pks and apply the order_by while turning back into a list of article objects.
+        articles = list(submission_models.Article.objects.filter(pk__in=article_list_unsorted).order_by(sort))
+
+    # just single keyword atm. but keyword is included in article_search.
+    elif keyword:
+        keyword_search = submission_models.Article.objects.filter(
+            keywords__word=keyword, 
+            journal=request.journal, 
+            stage=submission_models.STAGE_PUBLISHED, 
+            date_published__lte=timezone.now()
+        ).order_by(sort)
+        articles = [article for article in keyword_search]
+
+    # all keywords of published articles.
+    published_articles = submission_models.Article.objects.filter(
+        journal=request.journal,
+        stage=submission_models.STAGE_PUBLISHED
+        )
+    # based on published articles, return potential keywords
+    all_keywords = submission_models.Keyword.objects.filter(
```

Comment:
This can be a very large QuerySet for some of the journals we host, should be limited to the top keywords in the journal instead.

Reply:
I have a commit incoming that takes care of this by just providing the top 20 keywords of all published articles. But I just had a thought that perhaps they should be relative to the search term you just performed? 
If you perform a search then show the top keywords from the articles that were retrieved in this section, otherwise  show the top 20.

Reply:
![image](https://user-images.githubusercontent.com/1020384/54689738-7b25dd00-4af6-11e9-8cd1-bf7bf6bdfde0.png)
On our custom theme the keywords are returned (on the right) on the element that shows the articles so you can reference them there. Perhaps my above suggestion is redundant?

Reply:
I quite like this, we should definitely show the keywords with the article in our themes as well, so that the user can tell why the article was matched. As to the list of keywords, yes limiting by 20 should be fine.

Reply:
Would this PR require the change in `/templates/elements/journal/box_article.html`  to show keywords?  
that would alter how these appear in the `/articles` page which would be out of scope for this, no?
The CMESH theme uses a template: search_article so it's a separate styling from box_article.

Reply:
Yeah its fine, I can make the change to include the keywords myself once this PR has been merged.